### PR TITLE
FIX the ctrans module for fast reciprocal space conversions 

### PR DESCRIPTION
**Pull request by stuwilkins**
Do not merge yet ... work in progress but open for comments @tacaswell, @ericdill, @danielballan, @ambarb, @cmazzoli

### CODE REVIEW

**Comment by stuwilkins** [comment]
Working version now for the grinder. Testing needed. See:

**Comment by stuwilkins** [comment]
Working version now for the grinder. Testing needed. See:

https://gist.github.com/stuwilkins/c24c67cc254d02984add

**Comment by stuwilkins** [comment]
Ok this is mostly working. I think that @vivekthampy should now test if he can and then we are good to merge from my perspective.I ran the test notebooks and all ran OK. 

What about docs?

Attn @tacaswell @ericdill

**Comment by ericdill** [comment]
closed to force travis to use the new master branch that has lmfit pinned to 0.8.3 on travis

**Comment by stuwilkins** [comment]
Yay ... just passed the unit tests.

File: skxray/core/utils.py [hunk 1]
```diff
@@ -60,6 +60,7 @@
     except ImportError:
         ctrans = None
 
+ctrans.set_threads(1)
```
Comment (ericdill) [inline skxray/core/utils.py]: remove this default

**Comment by ericdill** [comment]
```
p(py3) ✘-1 ~/dev/python/scikit-xray [fix_ctrans|✚ 1…32]
14:05 $ python setup.py install
running install
running bdist_egg
running egg_info
writing top-level names to scikit_xray.egg-info/top_level.txt
writing requirements to scikit_xray.egg-info/requires.txt
writing dependency_links to scikit_xray.egg-info/dependency_links.txt
writing scikit_xray.egg-info/PKG-INFO
reading manifest file 'scikit_xray.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'scikit_xray.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-10.5-x86_64/egg
running install_lib
running build_py
creating build/lib.macosx-10.5-x86_64-3.4
creating build/lib.macosx-10.5-x86_64-3.4/skxray
copying skxray/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray
copying skxray/_version.py -> build/lib.macosx-10.5-x86_64-3.4/skxray
copying skxray/diffraction.py -> build/lib.macosx-10.5-x86_64-3.4/skxray
copying skxray/fluorescence.py -> build/lib.macosx-10.5-x86_64-3.4/skxray
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/arithmetic.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/calibration.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/cdi.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/correlation.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/dpc.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/feature.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/image.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/recip.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/roi.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/speckle.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/spectroscopy.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/stats.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
copying skxray/core/utils.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core
creating build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/avizo_io.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/binary.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/gsas_file_reader.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/net_cdf_io.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
copying skxray/io/save_powder_output.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io
creating build/lib.macosx-10.5-x86_64-3.4/skxray/testing
copying skxray/testing/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/testing
copying skxray/testing/decorators.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/testing
copying skxray/testing/noseclasses.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/testing
creating build/lib.macosx-10.5-x86_64-3.4/skxray/tests
copying skxray/tests/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/tests
copying skxray/tests/test_diffraction.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/tests
copying skxray/tests/test_fluorescence.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/tests
copying skxray/tests/test_openness.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/tests
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants
copying skxray/core/constants/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants
copying skxray/core/constants/basic.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants
copying skxray/core/constants/xrf.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants
copying skxray/core/constants/xrs.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/background.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/funcs.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/lineshapes.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/models.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
copying skxray/core/fitting/xrf_model.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_arithmetic.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_calibration.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_cdi.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_correlation.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_dpc.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_feature.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_image.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_recip.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_roi.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_speckle.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_spectroscopy.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_stats.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
copying skxray/core/tests/test_utils.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/tests
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
copying skxray/core/constants/tests/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
copying skxray/core/constants/tests/test_api.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
copying skxray/core/constants/tests/test_basic.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
copying skxray/core/constants/tests/test_xrf.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
copying skxray/core/constants/tests/test_xrs.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/tests
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/base
copying skxray/core/fitting/base/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/base
copying skxray/core/fitting/base/parameter_data.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/base
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/tests
copying skxray/core/fitting/tests/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/tests
copying skxray/core/fitting/tests/test_background.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/tests
copying skxray/core/fitting/tests/test_lineshapes.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/tests
copying skxray/core/fitting/tests/test_xrf_fit.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/fitting/tests
creating build/lib.macosx-10.5-x86_64-3.4/skxray/io/tests
copying skxray/io/tests/__init__.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io/tests
copying skxray/io/tests/test_api.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io/tests
copying skxray/io/tests/test_netCDF_io.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io/tests
copying skxray/io/tests/test_powder_output.py -> build/lib.macosx-10.5-x86_64-3.4/skxray/io/tests
creating build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/data
copying skxray/core/constants/data/AtomicConstants.dat -> build/lib.macosx-10.5-x86_64-3.4/skxray/core/constants/data
UPDATING build/lib.macosx-10.5-x86_64-3.4/skxray/_version.py
set build/lib.macosx-10.5-x86_64-3.4/skxray/_version.py to '0.0.5+35.g0028d51.dirty'
running build_ext
building 'ctrans' extension
gcc -fno-strict-aliasing -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/edill/mc/envs/py3/include -arch x86_64 -DUSE_THREADS -DDEBUG -I/Users/edill/mc/envs/py3/lib/python3.4/site-packages/numpy/core/include -I/Users/edill/mc/envs/py3/include/python3.4m -c src/ctrans.c -o build/temp.macosx-10.5-x86_64-3.4/src/ctrans.o
src/ctrans.c:55:10: fatal error: 'sys/sysinfo.h' file not found
#include <sys/sysinfo.h>
         ^
1 error generated.
error: command 'gcc' failed with exit status 1
```

Any idea why I can't install this on my mac?

**Comment by licode** [comment]
I think 'sysinfo.h' is linux specific header file for c/c++. Maybe you can use 'sysctl' in mac.

http://stackoverflow.com/questions/12523704/mac-os-x-equivalent-header-file-for-sysinfo-h-in-linux

**Comment by tacaswell** [comment]
These failures look like real regressions

```
======================================================================
FAIL: skxray.core.tests.test_utils.test_process_grid_std_err
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/mc/envs/testenv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-xray/scikit-xray/skxray/core/tests/test_utils.py", line 263, in test_process_grid_std_err
    np.std(np.arange(1, 6))/np.sqrt(5 - 1)))
  File "/home/travis/mc/envs/testenv/lib/python3.4/site-packages/numpy/testing/utils.py", line 739, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/travis/mc/envs/testenv/lib/python3.4/site-packages/numpy/testing/utils.py", line 665, in assert_array_compare
    raise AssertionError(msg)
nose.proxy.AssertionError: 
Arrays are not equal
(mismatch 100.0%)
 x: array([[[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],...
 y: array([[[ 0.707107,  0.707107,  0.707107,  0.707107,  0.707107,  0.707107,
          0.707107,  0.707107,  0.707107,  0.707107],
        [ 0.707107,  0.707107,  0.707107,  0.707107,  0.707107,  0.707107,...
-------------------- >> begin captured logging << --------------------
```

File: src/ctrans.c [hunk 2]
```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
...
+  unsigned long *n_grid = data->n_grid;
+  double *Mk = data->Mk;
+  double *Qk = data->Qk;
+  double *data_ptr = data->data;
+  double *dout = data->dout;
+  unsigned long *nout = data->nout;
+  unsigned long pos = 0;
+  
+  unsigned long i;
+
```
Comment (licode) [inline src/ctrans.c]: Need to add the following in this function if it is used in pthread_create.
```
#ifdef USE_THREADS
  pthread_exit(NULL);
#else
  return NULL;
```

File: src/ctrans.c [hunk 3]
```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
...
-
-  double pos_double[3];
-  double grid_len[3], grid_step[3];
-  int grid_pos[3];
-  int pos = 0;
-  unsigned long n_outside = 0;
+int c_grid3d(double *dout, unsigned long *nout, double *mout,
+             double *stderror, double *data, unsigned long *n_outside,
+             double *grid_start, double *grid_stop, unsigned long max_data,
+             unsigned long *n_grid, int norm){
```
Comment (licode) [inline src/ctrans.c]: It will be good if several lines of doc can be added to explain what this function does.

File: src/ctrans.c [hunk 4]
```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
...
-      return n_outside;
+ 
+  for(i=0;i<_n_threads;i++){
+    threadData[i].Mk = NULL;
+    threadData[i].Qk = NULL;
+    threadData[i].dout = NULL;
+    threadData[i].nout = NULL;
+  }
+
+  nthread = (unsigned long *)malloc(sizeof(unsigned long) * grid_size);
```
Comment (licode) [inline src/ctrans.c]: What is nthread used for?

File: src/ctrans.c [hunk 5]
```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
...
+  if(!nthread){
+    goto error;
+  }
+  for(i=0;i<grid_size;i++){
+    nthread[i] = 0;
+  }
+
+  stride = max_data / _n_threads;
+  for(i=0;i<_n_threads;i++){
+    threadData[i].Qk = (double*)malloc(sizeof(double) * grid_size);
```
Comment (licode) [inline src/ctrans.c]: There may be some memory leaking of those arrays unless they are freed at a top level.

**Comment by stuwilkins** [comment]
@tacaswell, this is actually a failure of how we calculate the std-dev. I did some debugging. The implemented way of doing this is to do the stddev on the fly ... The problem is that when you have a sparse grid in the map-reduce this fails because each bin. 

https://en.wikipedia.org/wiki/Standard_deviation#Rapid_calculation_methods

From this code Q_k fails if you have one value in a bin. It would be possible that if you had 32 values they could be distributed over all cores and then you get zero. It appear this is what is happening.

**Comment by stuwilkins** [comment]
@licode you are exactly correct. I noticed that and will fix ASAP.

File: src/ctrans.c (hunk 4 above)
Comment (stuwilkins) [inline src/ctrans.c]: It is a hangup from a previous code. I have removed it.

**Comment by stuwilkins** [comment]
@ericdill this should now be fixed on the mac... could you please try? Thanks.

S

**Comment by stuwilkins** [comment]
@tacaswell what should we do here? This is a failure of the math not of the code. Should we change the unit test?

File: src/ctrans.c [hunk 6]
```diff
@@ -475,155 +656,106 @@ unsigned long c_grid3d(double *dout, unsigned long *nout, double *standarderror,
...
+static PyObject* set_threads(PyObject *self, PyObject *args){
+  int threads;
+  if(!PyArg_ParseTuple(args, "i", &threads)){
+    return NULL;
+  }
+  if(threads > MAX_THREADS){
+    PyErr_SetString(PyExc_ValueError, "Requested number of threads > MAX_THREADS");
+    return NULL;
+  }
+  _n_threads = threads;
```
Comment (tacaswell) [inline src/ctrans.c]: This should be protected with `#ifdef USE_THREAD` to only allow `_n_threads` to be 1 if threading is not being used.

**Comment by tacaswell** [comment]
I would change it to keep track of the second moment which can be pooled to compute a single std rather than averaging the std's of each per-thread population.  

Are we in the regime where a single voxel's `s2` value will be large enough to hit the numerical issues?

**Comment by stuwilkins** [comment]
Hi @tacaswell what do you mean? I think that we are definitely in a regime where the s2 value will be large enough ... or you should not use the stddev calculation. It is really when N>> voxels

**Comment by stuwilkins** [comment]
@tacaswell , 

Your suggestion would require mutex locking between the second moment voxels or you would have a race condition on changing a single voxel from many threads. This would then need a mutex for each pixel. Or am I off base?????

**Comment by tacaswell** [comment]
Right now you are computing `Mk` and `Qk` per voxel per thread, I am suggesting computing a running sum of `val**2` instead.  

The reason to use the `Mk`/`Qk` scheme instead of compute the first and second moments is to avoid numerical issues with `s2 - s1 ** 2`.  How many hits-per-voxel are you expecting?

**Comment by ericdill** [comment]
@stuwilkins @tacaswell I remember the two of you were discussing something on the way to lunch. What is the status of this?

**Comment by stuwilkins** [comment]
Tested and working with threads. I think it is good to go

Best wishes,
Stuart


On Dec 9, 2015, at 5:09 PM, Eric Dill <notifications@github.com<mailto:notifications@github.com>> wrote:


@stuwilkins<https://github.com/stuwilkins> @tacaswell<https://github.com/tacaswell> I remember the two of you were discussing something on the way to lunch. What is the status of this?

**Comment by ericdill** [comment]
close/open to kick travis

**Comment by tacaswell** [comment]
I will do one more scan of this and merge today.

File: skxray/core/recip.py [hunk 7]
```diff
@@ -54,18 +54,12 @@
...
-        import ctrans
-    except ImportError:
-        ctrans = None
+from .ext import ctrans
 
 
 def process_to_q(setting_angles, detector_size, pixel_size,
                  calibrated_center, dist_sample, wavelength, ub,
-                 frame_mode=None):
+                 frame_mode=None, n_threads=0):
```
Comment (tacaswell) [inline skxray/core/recip.py]: I would prefer `None` as the 'use the default' value.

File: skxray/core/recip.py [hunk 8]
```diff
@@ -165,15 +164,15 @@ def process_to_q(setting_angles, detector_size, pixel_size,
                         ccd_cen=(calibrated_center),
                         dist=dist_sample,
                         wavelength=wavelength,
-                        UBinv=np.matrix(ub).I)
-    # **kwargs)
+                        UBinv=np.matrix(ub).I,
+                        n_threads=n_threads)
```
Comment (tacaswell) [inline skxray/core/recip.py]: put the

```python
if n_threads is None:
   n_threads = 0
```

just above this call.  Understand why you want int input to the function, but 'magic' int values is something we do not have to do in python.

Comment (tacaswell) [inline skxray/core/recip.py]: and, no, I will not get off your lawn

**Comment by ericdill** [comment]
@tacaswell bad review of this PR. :-1: 

The reason why there is a green check mark on this PR is because the tests were not even being run.  See [here] (https://travis-ci.org/scikit-beam/scikit-beam/jobs/96088377#L719).  I am trying to sort out how to get the tests to run with this modified code in #371

**Comment by tacaswell** [comment]
brilliant!

**Comment by stuwilkins** [comment]
@ericdill , that is really my fault not @tacaswell for not mentioning that I had problems with the c extension modules on travis and **I** didn't check that it ran the tests correctly. Thanks to @danielballan for a much more elegant solution!

**Comment by danielballan** [comment]
credit to @ericdill on the more elegant solution -- I just clicked the green button

**Comment by stuwilkins** [comment]
Ok ... thanks @ericdill !!!

**Pull request merged**